Small performance tweak: avoid List.map when computing maximum over list.

diesalbla opened the pull request:
The `Types` trait uses a method to compute the depth of a type. For composite types, this is based on the the maximum depth of its components. In some cases, this was using a call to `List.map(_.info)`, which creates a temporary list of types that is discarded afterwards.

- We add a generic `Depth.maximumBy` method, that computes the maximum over a list in constant space. It takes a function from a generic type to `Depth`.  We replace the loops to compute maximums with calls to this method.
- We invoke this method using different lambda functions, which saves the need to create the auxiliary lists.

diesalbla committed:
Type depth: avoid List.map when computing maximum over list.

The existing method to compute a `typeDepth` was using for some cases
a `map (_.info)` method, which creates a list that is then dropped.
We change the implementation of `maxDepth`:

- We add a `maximumBy` method to `Depth` companion object,
- We use lambda functions that bypass the `_.info`.

retronym commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -5130,20 +5130,8 @@ trait Types
-    }
-    loop(tps, Depth.Zero)
-  }
+  private[scala] def maxDepth(tps: List[Type]): Depth =
```
The comment above isn't needed now.

retronym commented on src/reflect/scala/reflect/internal/Depth.scala:
```diff
@@ -37,4 +37,9 @@ object Depth {
     else new Depth(depth)
   }
+
+  def maximumBy[A](xs: List[A], ff: A => Depth): Depth = {
```
Functions with value classes will incur boxing/unboxing. Introducing a SAM type might be better:

```scala
trait DepthFunction[A] { def apply(a: A): Depth }
object Depth {
def maximumBy[A](xs: List[A])(ff: DepthFunction): Depth = {
  var d = Depth(0)
  var rest = xs
  while (!rest.isEmpty) {
    d = ...
    rest = rest.tail
  }
  d
}
```

diesalbla committed:
Attend Code Review Comments

- Use a separate trait DepthFunction, to avoid boxing-unboxing.
- Remove unneeded comment.

retronym commented on src/reflect/scala/reflect/internal/Depth.scala:
```diff
@@ -37,4 +39,9 @@ object Depth {
+  def maximumBy[A](xs: List[A])(ff: DepthFunction[A]): Depth = {
+    def maxDepth(d: Depth, x: A) = d max ff(x)
+    xs.foldLeft(Zero)(maxDepth)
+  }
```
This needs to box the `Depth` value classes to work with the generic code of `List.foldLeft`. Use a `while` loop instead.

```
  public maximumBy(Lscala/collection/immutable/List;Lscala/reflect/internal/DepthFunction;)I
    // parameter final  xs
    // parameter final  ff
    ALOAD 1
    NEW scala/reflect/internal/Depth
    DUP
    ALOAD 0
    INVOKEVIRTUAL scala/reflect/internal/Depth$.Zero ()I
    INVOKESPECIAL scala/reflect/internal/Depth.<init> (I)V
    ALOAD 2
    INVOKEDYNAMIC apply(Lscala/reflect/internal/DepthFunction;)Lscala/Function2; [
      // handle kind 0x6 : INVOKESTATIC
      java/lang/invoke/LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;
      // arguments:
      (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      // handle kind 0x6 : INVOKESTATIC
      scala/reflect/internal/Depth$.$anonfun$maximumBy$1$adapted(Lscala/reflect/internal/DepthFunction;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      3,
      1,
      java.io.Serializable.class
    ]
    ASTORE 4
    ASTORE 3
    IFNONNULL L0
    ACONST_NULL
    ATHROW
   L0
    ALOAD 1
    ALOAD 3
    ALOAD 4
    INVOKESTATIC scala/collection/LinearSeqOps.foldLeft$ (Lscala/collection/LinearSeqOps;Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; (itf)
    CHECKCAST scala/reflect/internal/Depth
    INVOKEVIRTUAL scala/reflect/internal/Depth.depth ()I
    IRETURN
    MAXSTACK = 4
    MAXLOCALS = 5

  // access flags 0x1019
  public final static synthetic $anonfun$maximumBy$1(Lscala/reflect/internal/DepthFunction;ILjava/lang/Object;)I
    // parameter final  ff$1
    // parameter final  d
    // parameter final  x
    ILOAD 1
    ALOAD 2
    ALOAD 0
    INVOKESTATIC scala/reflect/internal/Depth$.maxDepth$1 (ILjava/lang/Object;Lscala/reflect/internal/DepthFunction;)I
    IRETURN
    MAXSTACK = 3
    MAXLOCALS = 3

  // access flags 0x1019
  public final static synthetic $anonfun$maximumBy$1$adapted(Lscala/reflect/internal/DepthFunction;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
    // parameter final  ff$1
    // parameter final  d
    // parameter final  x
    NEW scala/reflect/internal/Depth
    DUP
    ALOAD 0
    ALOAD 1
    CHECKCAST scala/reflect/internal/Depth
    INVOKEVIRTUAL scala/reflect/internal/Depth.depth ()I
    ALOAD 2
    INVOKESTATIC scala/reflect/internal/Depth$.$anonfun$maximumBy$1 (Lscala/reflect/internal/DepthFunction;ILjava/lang/Object;)I
    INVOKESPECIAL scala/reflect/internal/Depth.<init> (I)V
    ARETURN
    MAXSTACK = 5
    MAXLOCALS = 3
```

retronym reviewed (changes requested):
Requesting small change to `maximumBy` to avoid boxing.

diesalbla committed:
Code Review comments: use while loop to avoid boxing

Using the generic `List.foldLeft` was incurring the boxing penalty.
Using a while loop  should prevent that.

retronym commented:
Looks good now:

```
 // access flags 0x1
  // signature <A:Ljava/lang/Object;>(Lscala/collection/immutable/List<TA;>;Lscala/reflect/internal/DepthFunction<TA;>;)I
  // declaration: int maximumBy<A>(scala.collection.immutable.List<A>, scala.reflect.internal.DepthFunction<A>)
  public maximumBy(Lscala/collection/immutable/List;Lscala/reflect/internal/DepthFunction;)I
    // parameter final  xs
    // parameter final  ff
    ALOAD 1
    ASTORE 3
    ALOAD 0
    INVOKEVIRTUAL scala/reflect/internal/Depth$.Zero ()I
    ISTORE 4
   L0
    ALOAD 3
    INVOKEVIRTUAL scala/collection/immutable/List.isEmpty ()Z
    IFNE L1
    ALOAD 0
    ILOAD 4
    ALOAD 2
    ALOAD 3
    INVOKEVIRTUAL scala/collection/immutable/List.head ()Ljava/lang/Object;
    INVOKEINTERFACE scala/reflect/internal/DepthFunction.apply (Ljava/lang/Object;)I
    INVOKEVIRTUAL scala/reflect/internal/Depth$.max$extension (II)I
    ISTORE 4
    ALOAD 3
    INVOKEVIRTUAL scala/collection/immutable/List.tail ()Ljava/lang/Object;
    CHECKCAST scala/collection/immutable/List
    ASTORE 3
    GOTO L0
   L1
    ILOAD 4
    IRETURN
    MAXSTACK = 4
    MAXLOCALS = 5
```
